### PR TITLE
add sparc to the daily ci

### PR DESCRIPTION
**Pull request by zouboan**
## Summary
add sparc to the cibuild.sh
## Impact
sparc architecture's daily ci
## Testing
tested every command of function sparc-gcc-toolchain in cibuild.sh

### CODE REVIEW

**Comment by acassis** [comment]
Hi @zouboan seems it is failing to download/extract the toolchain:
```
> [nuttx-toolchain-sparc 1/1] RUN mkdir sparc-gaisler-elf-gcc &&   curl -L "https://www.gaisler.com/anonftp/bcc2/bin/bcc-2.1.0-gcc-linux64.tar.xz"   | tar -C sparc-gaisler-elf-gcc --strip-components 1 -xvJf:
------
executor failed running [/bin/sh -c mkdir sparc-gaisler-elf-gcc &&   curl -L "https://www.gaisler.com/anonftp/bcc2/bin/bcc-2.1.0-gcc-linux64.tar.xz"   | tar -C sparc-gaisler-elf-gcc --strip-components 1 -xvJf]: exit code: 64
Error: Process completed with exit code 1.
```

**Comment by zouboan** [comment]
the exact error is curl: (23) Failed writing body (0 != 7969)
&nbsp;
&nbsp;
------------------&nbsp;Original&nbsp;------------------
From: ***@***.***&gt;;
Date: &nbsp;Wed, Feb 2, 2022 09:08 PM
To: ***@***.***&gt;; 
Cc: ***@***.***&gt;; ***@***.***&gt;; 
Subject: &nbsp;Re: [apache/incubator-nuttx] add sparc to the daily ci (PR #5396)

&nbsp;



 
Hi @zouboan seems it is failing to download/extract the toolchain:
 &gt; [nuttx-toolchain-sparc 1/1] RUN mkdir sparc-gaisler-elf-gcc &amp;&amp;   curl -L "https://www.gaisler.com/anonftp/bcc2/bin/bcc-2.1.0-gcc-linux64.tar.xz"   | tar -C sparc-gaisler-elf-gcc --strip-components 1 -xvJf: ------ executor failed running [/bin/sh -c mkdir sparc-gaisler-elf-gcc &amp;&amp;   curl -L "https://www.gaisler.com/anonftp/bcc2/bin/bcc-2.1.0-gcc-linux64.tar.xz"   | tar -C sparc-gaisler-elf-gcc --strip-components 1 -xvJf]: exit code: 64 Error: Process completed with exit code 1.  
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by zouboan** [comment]
@acassis Could be the permissions issue of docker?
https://stackoverflow.com/questions/51044748/error-installing-docker-curl-23-failed-writing-body-0-13847

I have no idea of how to solve it!

**Comment by acassis** [comment]
> @acassis Could be the permissions issue of docker? https://stackoverflow.com/questions/51044748/error-installing-docker-curl-23-failed-writing-body-0-13847
> 
> I have no idea of how to solve it!

Yes, could be. But it is strange because for other similar commands it is working. Did you try wget instead of curl ?

File: tools/ci/docker/linux/Dockerfile [hunk 1]
```diff
@@ -97,6 +97,15 @@ RUN mkdir riscv64-unknown-elf-gcc && \
...
   | tar -C riscv64-unknown-elf-gcc --strip-components 1 -xz
 
+###############################################################################
+# Build image for tool required by SPARC builds
+###############################################################################
+FROM nuttx-toolchain-base AS nuttx-toolchain-sparc
+# Download the SPARC GCC toolchain prebuilt by Gaisler
+RUN mkdir sparc-gaisler-elf-gcc && \
+  curl -L "https://www.gaisler.com/anonftp/bcc2/bin/bcc-2.1.0-gcc-linux64.tar.xz" \
+  | tar -C sparc-gaisler-elf-gcc --strip-components 1 -xvJf
```
Comment (Ouss4) [inline tools/ci/docker/linux/Dockerfile:107]:
```suggestion
  | tar -C sparc-gaisler-elf-gcc --strip-components 1 -xJ
```

**Review by Ouss4 — COMMENTED** [review]
It was failing because the `f` option requires an extra argument (the file).  We don't need that as the file is piped. Also the `v` (verbose) option is not needed, it's better to have a cleaner output.

File: tools/ci/docker/linux/Dockerfile (hunk 1 above)
Comment (Ouss4) [reply tools/ci/docker/linux/Dockerfile:107]: Speaking of not needed, the manual says that `J` is not needed too.

Comment (zouboan) [reply tools/ci/docker/linux/Dockerfile]: thanks,it indeed solved problem,but J must need

**Comment by zouboan** [comment]
@acassis thanks,it seems that i have to do two pull request:
first: modify Dockerfile to install sparc toolchain in docker
second: modify tools/ci/testlist/other.dat to enable sparc ci
if i do these In one pull request,there's error in build xx3803:nsh can't find sparc-gaisler-elf-gcc,although sparc toolchain install success in Docker-Linux / push (pull_request)  checks

**Comment by Ouss4** [comment]
> @acassis thanks,it seems that i have to do two pull request: first: modify Dockerfile to install sparc toolchain in docker second: modify tools/ci/testlist/other.dat to enable sparc ci if i do these In one pull request,there's error in build xx3803:nsh can't find sparc-gaisler-elf-gcc,although sparc toolchain install success in Docker-Linux / push (pull_request) checks

Yes, you have to do that.  The reason is that the CI only builds the docker image in a pull request.  The push triggers only when it's merged to master.
So when the first PR is merged you can rebase the second one and it should pass.

**Comment by zouboan** [comment]
@Ouss4 Ok,thank you for clearing up my doubts!

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**